module/VuFind/sql/mysql.sql

```diff
@@ -227,17 +227,17 @@ CREATE TABLE `user` (
   `auth_method` varchar(50) DEFAULT NULL,
   `last_language` varchar(30) NOT NULL DEFAULT '',
   PRIMARY KEY (`id`),
-  UNIQUE KEY `username` (`username`),
-  UNIQUE KEY `cat_id` (`cat_id`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+  UNIQUE KEY `username` (`username`(190)),
```

Comment:
I'm worried about limiting UNIQUE keys to a specific amount of characters, because this might lead to side effects.
For example, if you have a user with a very long user name and another user would like to register with similar starting characters, this would not be possible because the DB would think  that the user already exists. So if there is a limit to a field you need a UNIQUE key on, the field length itself should be reduced to the max size of the key to avoid problems.
However, i can also understand that there will be existing installations for which we do not want to cause problems with existing users. And long user names should be the exception and not the regular case, so it might be worth taking that risk (or take the other way and remove the limit, but lose compatibility with old MySQL versions at the same time.)

Reply:
It's probably worth doing an experiment with creating 191-character usernames that only differ in the last character, and then running the database migration to see what happens. My suspicion is that the update will fail because MySQL won't let you create a unique key violation. If that is the case, I think we can probably just put a warning in the changelog so that people understand the situation if it occurs, and assume that this is extremely unlikely to happen in a real-world scenario. (Of course, we should also test that unique key violations are handled gracefully in the VuFind UI -- i.e. if a user tries to register with a duplicate key violation, they should get the "your username is already taken" type of message).

On the other hand, if the database upgrade DOESN'T fail in this scenario and instead creates some kind of problem/corruption, then we obviously need to change our approach and/or add some additional checks/validations to the upgrade logic.

In any case, I can probably find some time to test this next week... but if anyone else has time sooner, just let me know.

Reply:
I found some time today to experiment with this.

First, I created very long usernames and applied a shorter unique key to the user table. Duplicate keys initially caused a fatal error, but I added 36fd982 to handle them more gracefully. I imagine the check for duplicate key exceptions is a bit fragile (and perhaps region-specific), so I gave it a separate method to make it easier to locally override, or make more complex in the core.

Next, I tried creating multiple long usernames in a utf8mb3 database, and then applying the upgrade process to bring that up to utf8mb4. On my first attempt, it "just worked" with no problems, which didn't make sense... until I realized that the upgrade process was not applying unique key changes correctly. I added commit 1212660 to fix this problem. Now, when I upgrade the database and create a key violation, the upgrade process fails with a reasonably clear message:

![image](https://user-images.githubusercontent.com/309069/127335961-9bad01ee-5c0f-4425-af9e-c2b60518e607.png)

So, bottom line, I've fixed some problems and added some more robust error handling to the code, and I think these things are going to be useful regardless of our final decisions on other details... but the question is, what is our final decision?

Omitting the key length limit appeared to cause no problems for me when upgrading my database using MySQL 8.0.26. Is the limit related to a 5.x issue, or does the key length only become potentially problematic when usernames contain four-byte characters? Is there some other extreme scenario we should be testing?

Assuming that the key length change IS necessary for whatever reason, I'm not too concerned about it. I can't think of a realistic scenario where a username is longer than 190 characters AND it only differs from another existing username past character 190 (though please let me know if you think this really could happen). Even if that scenario happens, we can now handle it with some grace, either by treating it as a duplicate username or by reporting a problem during the upgrade process. If a user encounters the upgrade problem, they can either choose to edit mysql.sql to eliminate the key restriction (and deal with the possible consequences of that), or they can clean up their data and try again.

Any thoughts on how to proceed next? As you can see, I don't have particularly strong opinions one way or the other, but I'm also not as well-informed on all the details as others might be.